presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushAggregationThroughExchange.java

```diff
@@ -0,0 +1,408 @@
+import com.facebook.presto.sql.planner.plan.ChildReplacer;
+import com.facebook.presto.sql.planner.plan.ExchangeNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.PlanVisitor;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.facebook.presto.sql.tree.SymbolReference;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step.FINAL;
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step.PARTIAL;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.REPLICATE;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.util.stream.Collectors.toList;
+
+public class PushAggregationThroughExchange
+        implements PlanOptimizer
+{
+    private final Metadata metadata;
+
+    public PushAggregationThroughExchange(Metadata metadata)
+    {
+        this.metadata = metadata;
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        PlanWithProperty result = plan.accept(new Rewriter(symbolAllocator, idAllocator), Context.emptyContext());
+        return result.getNode();
+    }
+
+    private static class Context
+    {
+        private boolean throughExchange;
+        private Optional<AggregationNode> aggregationNode;
+
+        Context(Optional<AggregationNode> aggregationNode, boolean throughExchange)
+        {
+            this.aggregationNode = aggregationNode;
+            this.throughExchange = throughExchange;
+        }
+
+        public Optional<AggregationNode> getAggregationNode()
+        {
+            return aggregationNode;
+        }
+
+        public boolean isThroughExchange()
+        {
+            return throughExchange;
+        }
+
+        public static Context emptyContext()
+        {
+            return new Context(Optional.empty(), false);
+        }
+    }
+
+    private class Rewriter
+            extends PlanVisitor<Context, PlanWithProperty>
+    {
+        private final SymbolAllocator allocator;
+        private final PlanNodeIdAllocator idAllocator;
+
+        public Rewriter(SymbolAllocator allocator, PlanNodeIdAllocator idAllocator)
+        {
+            this.allocator = allocator;
+            this.idAllocator = idAllocator;
+        }
+
+        @Override
+        protected PlanWithProperty visitPlan(PlanNode node, Context context)
+        {
+            PlanNode newNode = node;
```

Comment:
it looks like you are trying to write your own plan traversal? why don't you just use like a SimplePlanRewriter? most of the other optimizers do that